pkg/istio-agent/xds-proxy.go

```diff
@@ -340,50 +375,78 @@ func buildUpstreamClientDialOpts(sa *Agent) ([]grpc.DialOption, error) {
 	// In these cases, while we fallback to mTLS to istiod using the provisioned certs
 	// it would be ideal to keep using token plus k8s ca certs for control plane communication
 	// as the intention behind provisioned certs on k8s pods is only for data plane comm.
-	if sa.secOpts.ProvCert == "" {
-		// only if running in k8s pod
-		dialOptions = append(dialOptions, grpc.WithPerRPCCredentials(oauth.TokenSource{&fileTokenSource{
+	if !sa.secOpts.FileMountedCerts && sa.secOpts.ProvCert == "" {
+		dialOptions = append(dialOptions, grpc.WithPerRPCCredentials(oauth.TokenSource{TokenSource: &fileTokenSource{
 			sa.secOpts.JWTPath,
 			time.Second * 300,
 		}}))
 	}
 	return dialOptions, nil
 }
 
+// initCertificateWatches sets up  watches for the certs and resets upstream if they change.
+func (p *XdsProxy) initCertificateWatches(agent *Agent, stop <-chan struct{}) error {
+	keyFile, certFile := p.getCertKeyPaths(agent)
+	rootCert := agent.FindRootCAForXDS()
+	for _, file := range []string{rootCert, certFile, keyFile} {
+		log.Infof("adding watcher for certificate %s", file)
+		if err := p.fileWatcher.Add(file); err != nil {
+			return fmt.Errorf("could not watch %v: %v", file, err)
+		}
+	}
+	go func() {
+		var keyCertTimerC <-chan time.Time
+		for {
+			select {
+			case <-keyCertTimerC:
+				keyCertTimerC = nil
+				proxyLog.Info("xds connection certificates have changed, resetting the upstream connection")
+				// Close upstream connection.
+				p.resetChan <- struct{}{}
+			case <-p.fileWatcher.Events(certFile):
+				if keyCertTimerC == nil {
+					keyCertTimerC = time.After(watchDebounceDelay)
+				}
+			case <-p.fileWatcher.Events(keyFile):
+				if keyCertTimerC == nil {
+					keyCertTimerC = time.After(watchDebounceDelay)
+				}
+			case <-stop:
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
 // Returns the TLS option to use when talking to Istiod
 // If provisioned cert is set, it will return a mTLS related config
 // Else it will return a one-way TLS related config with the assumption
 // that the consumer code will use tokens to authenticate the upstream.
-func getTLSDialOption(agent *Agent) (grpc.DialOption, error) {
-	var certPool *x509.CertPool
-	var err error
-	var rootCert []byte
-	xdsCACert := agent.FindRootCAForXDS()
-	rootCert, err = ioutil.ReadFile(xdsCACert)
+func (p *XdsProxy) getTLSDialOption(agent *Agent) (grpc.DialOption, error) {
+
+	rootCert, err := p.getRootCertificate(agent)
 	if err != nil {
 		return nil, err
 	}
 
-	certPool = x509.NewCertPool()
-	ok := certPool.AppendCertsFromPEM(rootCert)
-	if !ok {
-		return nil, fmt.Errorf("failed to create TLS dial option with root certificates")
-	}
-
-	var certificate tls.Certificate
 	config := tls.Config{
-		Certificates: []tls.Certificate{certificate},
 		GetClientCertificate: func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
-			if agent.secOpts.ProvCert != "" {
+			var certificate tls.Certificate
```

Comment:
I don't understand why we need to reset explicitly. I thought that GetClientCertificate is called on every new connection already? Even if the cert changes, our existing connection will have already done the TLS handshake and should be fine? This code seems to drift from CSR code

Reply:
Yes. Existing connections is fine. The new connection might only happen on the next disconnect either forceful disconnect by Istiod or because of some network disconnect. Assuming that we want to do intelligent load balancing at proxy in future, the forceful disconnect might not be there. 

So do we want the new certs only to be picked on new connections (which is not in our control when they might happen)?  Comparing with today's functionality with out xds proxy, the certs would be reloaded when they change via SDS? I am trying to match that behaviour here.

But if we think, it is better to not watch and let the new certs take affect only on new connection - I am fine with that.